kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaRestoreService.java

```diff
@@ -0,0 +1,160 @@
+          "Replica restore task already scheduled, will run in {} ms",
+          pendingTask.getDelay(TimeUnit.MILLISECONDS));
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting replica restore service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closing replica restore service");
+    executorService.shutdownNow();
+    LOG.info("Closed replica restore service");
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        replicaRestoreServiceConfig.getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  /** Queues snapshots to be restored in the future */
+  public synchronized void queueSnapshotsForRestoration(List<SnapshotMetadata> snapshotsToRestore)
+      throws Exception {
+    if (snapshotsToRestore.size() >= replicaRestoreServiceConfig.getMaxReplicasPerRequest()) {
+      throw new SizeLimitExceededException();
+    }
+    queue.addAll(snapshotsToRestore);
+    LOG.info("Current size of Snapshot restoration queue: " + queue.size());
+    runOneIteration();
+  }
+
+  /** Drains the current queue and creates replicas as required. Called by scheduler. */
+  private void restoreQueuedSnapshots() {
+    if (queue.isEmpty()) {
+      return;
+    }
+
+    Timer.Sample restoreSnapshotsTimer = Timer.start(meterRegistry);
+
+    List<SnapshotMetadata> snapshotsToRestore = new ArrayList<>();
+    Set<String> createdReplicas = new HashSet<>();
+
+    for (ReplicaMetadata replicaMetadata : replicaMetadataStore.getCached()) {
+      createdReplicas.add(replicaMetadata.snapshotId);
+    }
+
+    queue.drainTo(snapshotsToRestore);
+
+    for (SnapshotMetadata snapshotMetadata : snapshotsToRestore) {
+      try {
+        restoreOrSkipSnapshot(snapshotMetadata, createdReplicas);
+        createdReplicas.add(snapshotMetadata.snapshotId);
+      } catch (InterruptedException e) {
+        LOG.error("Something went wrong dequeueing snapshot ID {}", snapshotMetadata.snapshotId, e);
+        failedReplicas.increment();
+      }
+    }
+    restoreSnapshotsTimer.stop(restoreTimer);
+  }
+
+  /** Creates replica from given snapshot if its ID doesn't already exist in createdReplicas */
+  private void restoreOrSkipSnapshot(SnapshotMetadata snapshot, Set<String> createdReplicas)
+      throws InterruptedException {
+    if (!createdReplicas.contains(snapshot.snapshotId)) {
+      LOG.info("Restoring replica with ID {}", snapshot.snapshotId);
+
+      try {
+        replicaMetadataStore.createSync(
+            replicaMetadataFromSnapshotId(
+                snapshot.snapshotId,
+                Instant.now()
+                    .plus(
+                        replicaRestoreServiceConfig.getReplicaLifetimeMins(), ChronoUnit.MINUTES)));
+      } catch (Exception e) {
+        LOG.error("Error restoring replica for snapshot {}", snapshot.snapshotId);
```

Comment:
We should include the error here - anytime there's an error we should always pass the exception